lib/EnzymeCore/src/rules.jl

```diff
@@ -208,18 +208,36 @@ function isapplicable(@nospecialize(f), @nospecialize(TT);
         matches = result
     end
     fullmatch = Core.Compiler._any(match::Core.MethodMatch->match.fully_covers, matches)
-    if caller !== nothing
-        fullmatch || add_mt_backedge!(caller, mt, sig)
+    if !fullmatch
+        if caller isa Core.MethodInstance
+            add_mt_backedge!(caller, mt, sig)
+        elseif caller isa Core.Compiler.MethodLookupResult
+            for j = 1:Core.Compiler.length(caller)
+                cmatch = Core.Compiler.getindex(caller, j)::Core.MethodMatch
+                cspec = Core.Compiler.specialize_method(cmatch)::Core.MethodInstance
+                add_mt_backedge!(cspec, mt, sig)
+            end
+        end
     end
```

Comment:
Can you explain why this code became necessary?

Reply:
Yeah so basically it was required to fix a bug in the rules tests.

in particular if a function was defined first without a rule, absint would succeed and let’s say inline the function. The only other time we add a back edge is when we take the actual generated llvm functions, which only exist for non Julia inlined code. So we would never add a back edge for the rule currently without a rule.

we would then go and add a rule, but that wouldn’t cause an invalidation and thus end up with the wrong code

Reply:
Yeah, that was known as a bug...

Julia solves this, by having a new method definition invalidate the method-table. We could change our rule definitions to a macro form that does that.